# Problema

Predecir el tipo de flor

# El set de datos

Número de Instancias:

El conjunto de datos Iris consta de 150 muestras de flores Iris con 4 características cada una.

Características:

* Longitud del sépalo (sepal length)
* Ancho del sépalo (sepal width)
* Longitud del pétalo (petal length)
* Ancho del pétalo (petal width)

Clases o Etiquetas:

* Setosa
* Versicolor
* Virginica



In [1]:
# imports
from sklearn.datasets import load_iris
import numpy as np
import pandas as pd

In [2]:
# Cargamos los datos de iris
# Load the Iris dataset
iris = load_iris()

# Create a DataFrame
iris_df = pd.DataFrame(data=iris.data, columns=iris.feature_names)

# Add the target variable (species) to the DataFrame
iris_df['target'] = iris.target
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

# Objetivo

Generar un model de clasificación capaz de predecir la clase de flor en función de las carácterísticas del dataset

* Aplicar las técnicas oportunas de procesamiento de datos

* Valorar diferentes modelos de clasificación

* Comparación entre modelos

* Ensemble

* Métricas

* Conclusiones finales

### Random Forest


1. Preparación de los Datos
Primero, se cargaron las librerías necesarias y se prepararon los datos, dividiéndolos en conjuntos de entrenamiento y prueba. Esto se hace para poder entrenar el modelo con un conjunto de datos y luego evaluar su rendimiento con otro conjunto, asegurando así que el modelo pueda generalizar bien a nuevos datos.

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Cargar el dataset Iris y convertirlo en un DataFrame
iris_data = load_iris()
iris_df = pd.DataFrame(data=iris_data['data'], columns=iris_data['feature_names'])
iris_df['target'] = iris_data['target']

# Separar las características (X) de las etiquetas (y)
X = iris_df.drop('target', axis=1)
y = iris_df['target']

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


2. Entrenamiento del Modelo
A continuación, se entrenó el modelo Random Forest con el conjunto de entrenamiento. Se eligió Random Forest por su eficacia en tareas de clasificación, especialmente con datasets como el de Iris.

In [11]:
# Entrenar el modelo Random Forest
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)
rf_classifier.fit(X_train, y_train)


RandomForestClassifier(random_state=42)

3. Evaluación del Modelo
Finalmente, se evaluó el rendimiento del modelo utilizando el conjunto de prueba. Se calcularon la precisión del modelo, el informe de clasificación y la matriz de confusión para entender mejor su rendimiento.

In [12]:
# Realizar predicciones con el conjunto de prueba
y_pred = rf_classifier.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)

# Generar el informe de clasificación
classification_rep = classification_report(y_test, y_pred)

# Generar la matriz de confusión
confusion_mat = confusion_matrix(y_test, y_pred)

print(f"Accuracy: {accuracy}")
print("\nClassification Report:\n", classification_rep)
print("\nConfusion Matrix:\n", confusion_mat)


Accuracy: 1.0

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      1.00      1.00        13
           2       1.00      1.00      1.00        13

    accuracy                           1.00        45
   macro avg       1.00      1.00      1.00        45
weighted avg       1.00      1.00      1.00        45


Confusion Matrix:
 [[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]


In [13]:
from sklearn.model_selection import cross_val_score

# Validación cruzada con 5 pliegues
cv_scores = cross_val_score(rf_classifier, X, y, cv=5)

print(f"Cross-validation scores: {cv_scores}")
print(f"Mean CV accuracy: {np.mean(cv_scores)}")


Cross-validation scores: [0.96666667 0.96666667 0.93333333 0.96666667 1.        ]
Mean CV accuracy: 0.9666666666666668


### KNN

In [6]:
from sklearn.neighbors import KNeighborsClassifier

# Configurar el modelo KNN
knn_classifier = KNeighborsClassifier(n_neighbors=5)

# Entrenar el modelo KNN
knn_classifier.fit(X_train, y_train)

# Realizar predicciones con el conjunto de prueba usando KNN
y_pred_knn = knn_classifier.predict(X_test)

# Evaluar el modelo KNN
accuracy_knn = accuracy_score(y_test, y_pred_knn)
classification_rep_knn = classification_report(y_test, y_pred_knn)
confusion_mat_knn = confusion_matrix(y_test, y_pred_knn)


print(f"Accuracy: {accuracy_knn}")
print("\nClassification Report:\n", classification_rep_knn)
print("\nConfusion Matrix:\n", confusion_mat_knn)

Accuracy: 1.0

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      1.00      1.00        13
           2       1.00      1.00      1.00        13

    accuracy                           1.00        45
   macro avg       1.00      1.00      1.00        45
weighted avg       1.00      1.00      1.00        45


Confusion Matrix:
 [[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]


In [14]:
from sklearn.model_selection import cross_val_score

# Validación cruzada con 5 pliegues
cv_scores = cross_val_score(knn_classifier, X, y, cv=5)

print(f"Cross-validation scores: {cv_scores}")
print(f"Mean CV accuracy: {np.mean(cv_scores)}")


Cross-validation scores: [0.96666667 1.         0.93333333 0.96666667 1.        ]
Mean CV accuracy: 0.9733333333333334


### XGBoost

In [16]:
from xgboost import XGBClassifier

# Configurar el modelo XGBoost
xgb_classifier = XGBClassifier(use_label_encoder=False, eval_metric='mlogloss')

# Entrenar el modelo XGBoost
xgb_classifier.fit(X_train, y_train)

# Realizar predicciones con el conjunto de prueba usando XGBoost
y_pred_xgb = xgb_classifier.predict(X_test)

# Evaluar el modelo XGBoost
accuracy_xgb = accuracy_score(y_test, y_pred_xgb)
classification_rep_xgb = classification_report(y_test, y_pred_xgb)
confusion_mat_xgb = confusion_matrix(y_test, y_pred_xgb)

print(f"Accuracy: {accuracy_xgb}")
print("\nClassification Report:\n", classification_rep_xgb)
print("\nConfusion Matrix:\n", confusion_mat_xgb)

Accuracy: 1.0

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      1.00      1.00        13
           2       1.00      1.00      1.00        13

    accuracy                           1.00        45
   macro avg       1.00      1.00      1.00        45
weighted avg       1.00      1.00      1.00        45


Confusion Matrix:
 [[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]


In [15]:
from sklearn.model_selection import cross_val_score

# Validación cruzada con 5 pliegues
cv_scores = cross_val_score(xgb_classifier, X, y, cv=5)

print(f"Cross-validation scores: {cv_scores}")
print(f"Mean CV accuracy: {np.mean(cv_scores)}")


Cross-validation scores: [0.96666667 0.96666667 0.93333333 0.93333333 0.96666667]
Mean CV accuracy: 0.9533333333333334


### Ensemble

In [8]:
from sklearn.ensemble import VotingClassifier

# Configurar el ensemble con votación dura
ensemble = VotingClassifier(estimators=[
    ('random_forest', rf_classifier),
    ('knn', knn_classifier)
], voting='hard')

# Entrenar el ensemble
ensemble.fit(X_train, y_train)

# Realizar predicciones con el conjunto de prueba usando el ensemble
y_pred_ensemble = ensemble.predict(X_test)

# Evaluar el ensemble
accuracy_ensemble = accuracy_score(y_test, y_pred_ensemble)
classification_rep_ensemble = classification_report(y_test, y_pred_ensemble)
confusion_mat_ensemble = confusion_matrix(y_test, y_pred_ensemble)

(accuracy_ensemble, classification_rep_ensemble, confusion_mat_ensemble)


(1.0,
 '              precision    recall  f1-score   support\n\n           0       1.00      1.00      1.00        19\n           1       1.00      1.00      1.00        13\n           2       1.00      1.00      1.00        13\n\n    accuracy                           1.00        45\n   macro avg       1.00      1.00      1.00        45\nweighted avg       1.00      1.00      1.00        45\n',
 array([[19,  0,  0],
        [ 0, 13,  0],
        [ 0,  0, 13]]))

# Conclusiones

El dataset iris es realmente sencillo de predecir ya que probablemente este bastante limpio y no haya muestras muy raras o clases muy desbalanceadas. Todos los modelos nos dan un accuracy cercano a 100%. Aún asi, realizando cross_validation vemos que el clasificador KNN funciona mejor de media, con un 97% de accuracy.